### ISSUE TITLE
CFP: ClusterMesh global service cluster affinity

### ISSUE THREAD

**Issue by dboslee** [cilium/cilium#29200]
## Cilium Feature Proposal

**Is your proposed feature related to a problem?**

Hey cilium folks thanks for the work you do on this project 🙌.

We use clustermesh/global services pretty heavily in our multi cluster environment. Our architecture is basically a cluster per region across several regions. Some of our services run in a cluster or two but not the others. We use global services in this case to reach those services when they aren't running in a given cluster. 

Time and time again when deploying services to a subset of clusters I wish I had more control than simply `service.cilium.io/affinity: local|remote`. The only options are to route to local endpoints, remote endpoints, or all endpoints.

For low latency or for active/passive configurations this is not enough when you have more than 2 clusters.

As an example lets say we have endpoints in our clusters in us-east-1 and us-west-2 for the global service `backend`. When accessing this service from a cluster in `eu-central-1` there is no choice but to load balance requests across the endpoints in the other two regions. Ideally we could prefer to route traffic to the `us-east-1` cluster for lower latency.

We have worked around this in some cases in order to route traffic to a specific cluster by creating a `backend-us-east-1` global service that only has endpoints in the `us-east-1` cluster. The downside to this approach is that there is no failover to endpoints in other clusters if the endpoints in`us-east-1` goes down.

**Describe the feature you'd like**

I would like a way to specify cluster affinity. This would behave similar to the existing affinity annotation however instead you could specify which cluster to have affinity towards. When the endpoints in that cluster are not available the service will fall back to load balancing across any other available endpoints.

This could look something like `service.cilium.io/cluster-affinity: cluster-name`.

**Comment by dboslee** [comment]
Yes I could contribute this if its something you all want to support. We had some additional ideas around what the api could look like for this as well.

Cluster affinity is probably the simplest most straight forward approach but an alternative approach could be to use label affinity. Something like...

`service.cilium.io/label-affinity/<label>: <value>`

This would work as a label selector allowing someone to match a subset of remote cluster services not just a single cluster like cluster affinity. Again falling back to all endpoints when the subset of services matched have no endpoints.

In a similar vein, some sort of priority could be given to clusters or labels to enforce a failover ordering. The API for this could get pretty messy though.

Anyways, I think either approach would be a great enhancement to global services. Let me know what you think.

**Comment by giorio94** [comment]
Hey @dboslee, thanks for your interest!

You proposal seems interesting to me, and it also fits nicely with the existing architecture. I'd personally prefer to keep the API as simple as possible, especially given that we need to use annotations to express it. In this respect, I think that your initial proposal looks pretty straightforward. I'm not opposed to generalizing it a bit, though, like:

```
service.cilium.io/affinity: cluster
service.cilium.io/cluster-affinity: cluster1,cluster2,...
```

Where setting `service.cilium.io/affinity` to cluster enabes this new mode, to avoid conflicting with the existing ones if both annotations are present, and `service.cilium.io/cluster-affinity` lists a set of preferred clusters. I'm not yet totally convinced whether it could be useful if the list implied a priority order (i.e., use the endpoints from cluster2 only if none are present in cluster1 vs endpoints from cluster1 and cluster2 are preferred), but avoiding that would simplify the design for sure.

I'd prefer not using the label affinity instead, both because there's no per-cluster labels concept, and due to the additional dynamism that would need to be accounted for. Matching on node labels seems also out of scope (as not clustermesh related), and it would come with non-trivial implications due to the need for propagating information about the node hosting each backend, and merging that with node labels.

**Comment by dboslee** [comment]
> I'd personally prefer to keep the API as simple as possible

👍 totally understand. 
<br>

> service.cilium.io/affinity: cluster

Great idea this makes sense to me.
<br>

> I'm not yet totally convinced whether it could be useful if the list implied a priority order

I see pros/cons for each approach, either of which would help our use case.

I think priority is useful specifically for latency. You could define which cluster to go to in order of latency. For example if you are connecting to a global service from `eu-west-1` you may prefer to failover to other clusters/regions in the order of `eu-central-1 (20ms)` `us-east-1 (70ms)` `us-west-2 (120ms)`.
<br>

> I'd prefer not using the label affinity instead, both because there's no per-cluster labels concept, and due to the additional dynamism that would need to be accounted for. 

Just to clarify I was referring to matching the labels present on the services themselves. My understanding is this is already present in the ClusterService here: https://github.com/cilium/cilium/blob/70ae8d0ef536de807aab849291e5a68758cb8d4d/pkg/service/store/store.go#L66-L67

But agreed even if it works it would still be a weird/complex api and would be happy to see the cluster affinity move forward.

**Comment by giorio94** [comment]
> I see pros/cons for each approach

Exactly. Ideally, we could support both approaches, exposing them through a different `service.cilium.io/affinity` value. Maybe something like:

```
service.cilium.io/affinity: cluster # any cluster specified in the cluster-affinity label is preferred
service.cilium.io/affinity: cluster-priority # the clusters specified in the cluster-affinity label are preferred in priority order
```

Which also simplifies implementing this feature in an incremental way, starting from the first alternative which is easier.

> Just to clarify I was referring to matching the labels present on the services themselves

Oh, then I misunderstood your proposal. But yeah, I'd still prefer to avoid label affinities, to keep it simple. 

> Yes I could contribute this if its something you all want to support

Great to hear. Feel free to reach out if you need any help. In any case, a good starting point would be the [`correlateEndpoints`](https://github.com/cilium/cilium/blob/076653691d0d15b4449d6e2791bdbaea5298ed5f/pkg/k8s/service_cache.go#L486) function, which allows to configure whether a given endpoint is preferred or not (and is where the current local/remote affinities are configured).

**Comment by dboslee** [comment]
not stale. I will be taking a look at implementing this sometime in the next month.

**Comment by giorio94** [comment]
I'm personally not. Feel free to go ahead, thanks!

**Comment by dboslee** [comment]
Feel free to take it. I got busy with other things unfortunately.

I had this stashed away in git from months ago. May be helpful as a jumping off point. https://github.com/cilium/cilium/compare/main...dboslee:cilium:david/cluster-affinity?expand=1
